src/serac/physics/materials/tests/nonlinear_J2_material.cpp

```diff
@@ -77,6 +77,166 @@ TEST(NonlinearJ2Material, Uniaxial)
   }
 };
 
+TEST(FiniteDeformationNonlinearJ2Material, Uniaxial)
+{
+  /* Log strain J2 plasticity has the nice feature that the exact uniaxial stress solution from
+     small strain plasticity are applicable, if you replace the lineasr strain with log strain
+     and use the Kirchhoff stress as the output.
+  */
+  double                             E       = 1.0;
+  double                             nu      = 0.25;
+  double                             sigma_y = 0.01;
+  double                             Hi      = E / 100.0;
+  double                             eps0    = sigma_y / Hi;
+  double                             n       = 1;
+  solid_mechanics::PowerLawHardening hardening{.sigma_y = sigma_y, .n = n, .eps0 = eps0};
+  solid_mechanics::J2FiniteDeformationNonlinear<decltype(hardening)> material{
+      .E = E, .nu = nu, .hardening = hardening, .density = 1.0};
+
+  auto internal_state   = solid_mechanics::J2FiniteDeformationNonlinear<decltype(hardening)>::State{};
+  auto strain           = [=](double t) { return sigma_y / E * t; };
+  auto response_history = uniaxial_stress_test(2.0, 4, material, internal_state, strain);
+
+  auto stress_exact = [=](double epsilon) {
+    return epsilon < sigma_y / E ? E * epsilon : E / (E + Hi) * (sigma_y + Hi * epsilon);
+  };
+  auto plastic_strain_exact = [=](double epsilon) {
+    return epsilon < sigma_y / E ? 0.0 : (E * epsilon - sigma_y) / (E + Hi);
+  };
+
+  for (auto r : response_history) {
+    double J  = detApIm1(get<1>(r)) + 1;
+    double e  = std::log1p(get<1>(r)[0][0]);       // log strain
+    double s  = get<2>(r)[0][0] * J;               // Kirchhoff stress
+    double pe = -std::log(get<3>(r).Fpinv[0][0]);  // plastic strain
+    ASSERT_NEAR(s, stress_exact(e), 1e-6 * std::abs(stress_exact(e)));
+    ASSERT_NEAR(pe, plastic_strain_exact(e), 1e-6 * std::abs(plastic_strain_exact(e)));
+  }
+};
+
+TEST(FiniteDeformationNonlinearJ2Material, DerivativeCorrectness)
+{
+  // This constitutive function is non-differentiable at the yield point,
+  // but should be differentiable everywhere else.
+  // The elastic response is trivial. We want to check the plastic reponse
+  // and make sure the derivative propagates correctly through the nonlinear
+  // solve.
+
+  // parameters
+  double E       = 200.0e9;
+  double nu      = 0.25;
+  double sigma_y = 350e6;
+  double eps0    = sigma_y / E;
+  double n       = 3;
+
+  // hardening model
+  solid_mechanics::PowerLawHardening hardening{.sigma_y = sigma_y, .n = n, .eps0 = eps0};
+
+  // material model
+  solid_mechanics::J2FiniteDeformationNonlinear<decltype(hardening)> material{
```

Comment:
most of the `decltype(hardening)` in these tests can be omitted if we add a Class Template Argument Deduction guide (see https://godbolt.org/z/P1q1v3hPG, although CTAD doesn't work with the designated initializer syntax `{.foo = foo, .bar = bar, ...}`).

Alternatively, some `using` statements like

```cpp
using hardening_t = solid_mechanics::PowerLawHardening;
using material_model_t = solid_mechanics::J2FiniteDeformationNonlinear< hardening_t >;
```

at the top of the file could make some of the declarations in it a little cleaner

Reply:
Thanks, CTAD was new to me, so I learned something. For this case, I chose the second route, because we want the materials to be simple for C++ laypeople to implement, and I don't want to give users the impression that CTAD is a necessary ingredient.